### PR TITLE
Fallback refactor + unit tests

### PR DESCRIPTION
**Pull request by tomaszmichalak**
Fallback mechanism refactor: remove internal cache, fixed the flow, added Unit tests for all cases.

## Description
Refactor.

## Motivation and Context
This module was unstable. We added unit tests and fixed implementation.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: knotx-core/src/test/java/io/knotx/fallback/FragmentFallbackHandlerTest.java [hunk 1]
```diff
@@ -15,90 +15,117 @@
...
+    new FragmentFallbackHandler(new FragmentFallbackHandlerOptions()).handle(mockContext);
+
+    // then
+    FragmentsContext result = mockContext.get(FragmentsContext.KEY);
+    Assert.assertEquals(1, result.getFragments().size());
+    Assert.assertEquals(actualBody, result.getFragments().get(0).getBody());
+  }
+
+  @Test
+  public void callHandlerWithFailedFragment_expectDefaultFallback() {
```
Comment (Skejven) [inline knotx-core/src/test/java/io/knotx/fallback/FragmentFallbackHandlerTest.java]: `..._whenNoFallbackConfigured...`?

File: knotx-core/src/test/java/io/knotx/fallback/FragmentFallbackHandlerTest.java [hunk 2]
```diff
@@ -15,90 +15,117 @@
...
+
+    // then
+    FragmentsContext result = mockContext.get(FragmentsContext.KEY);
+    Assert.assertEquals(1, result.getFragments().size());
+    Assert.assertEquals(FallbackConstants.EMPTY_FALLBACK_VALUE,
+        result.getFragments().get(0).getBody());
+  }
+
+  @Test
+  public void callHandlerWithFailedFragmentWithCustomMetadataFallback_expectDefaultFallback() {
```
Comment (Skejven) [inline knotx-core/src/test/java/io/knotx/fallback/FragmentFallbackHandlerTest.java]: `...expectCustomFallbackApplied`?

File: knotx-core/src/test/java/io/knotx/fallback/FragmentFallbackHandlerTest.java [hunk 3]
```diff
@@ -15,90 +15,117 @@
...
+    new FragmentFallbackHandler(options).handle(mockContext);
+
+    // then
+    FragmentsContext result = mockContext.get(FragmentsContext.KEY);
+    Assert.assertEquals(1, result.getFragments().size());
+    Assert.assertEquals(CUSTOM_FALLBACK_MESSAGE, result.getFragments().get(0).getBody());
+  }
+
+  @Test
+  public void callHandlerWithFailedFragmentWithCustomFragmentFallback_expectDefaultFallback() {
```
Comment (Skejven) [inline knotx-core/src/test/java/io/knotx/fallback/FragmentFallbackHandlerTest.java]: Shouldn't there be expected `customFallback`?

File: knotx-core/src/main/java/io/knotx/fallback/FallbackConstants.java [hunk 4]
```diff
@@ -15,12 +15,14 @@
  */
 package io.knotx.fallback;
 
-public class FragmentFallbackConstants {
+public class FallbackConstants {
```
[the remark this answers is not in the document]
Comment (tomaszmichalak) [reply knotx-core/src/main/java/io/knotx/fallback/FallbackConstants.java]: done

**Pull request merged**